### ISSUE TITLE
New rule: no-flowmax-in-forwardref?

### ISSUE THREAD

**Issue by helixbass** [helixbass/eslint-plugin-ad-hok#2]
I just had to debug for a while to figure out that having a "magic" `flowMax()` defined "inline" inside a `React.forwardRef()`-component callback was resulting in the recreation of the component (for the post-magic part of the `flowMax()`) on every render, which makes it remount on every render because React thinks it's a different component

So could presumably catch this case by looking for `flowMax()` inside `forwardRef()`?

I don't know if I'd worry about false positives for non-React `flowMax()`'s?

I don't think you could generalize to "flag any `flowMax()` not at the top level of the module"?

**Comment by helixbass** [comment]
@peterstuart here are some cases that should fail (ie trigger ESLint):
```
export default forwardRef((props, ref) =>
  callWith({...props, ref})(
    flowMax(
      addProps({foo: 'bar'}),
      ({foo, ref}) =>
        <div ref={ref}>{foo}</div>
    )
  )
)
```
```
const Foo = React.forwardRef((props, forwardedRef) => {
  const InnerFoo = flowMax(
    ({bar, forwardedRef}) =>
      <div ref={forwardedRef}>{bar}</div>
  )
  return <InnerFoo {...props} forwardedRef={forwardedRef} />
})
```
and some passing cases:
```
flowMax()
```
```
const HoistedComponent = flowMax(
  addWrapper(({render, forwardedRef}) => <div ref={forwardedRef}>{render()}</div>),
  ({foo}) => <span>{foo}</span>
)

const Component = React.forwardRef((props, ref) =>
  <HoistedComponent {...props} forwardedRef={ref} />
)
```
```
const Foo = forwardRef((props, ref) =>
  flow(
    addProps(({foo}) => ({bar: foo + 1})),
    ({bar, ref}) => <div ref={ref}>{bar}</div>
  )({...props, ref})
)
```

**Comment by helixbass** [comment]
Fixed by #16
